allow-multiple-versions/src/multiple_versions.rs

```diff
@@ -0,0 +1,315 @@
+            Package::Proxi { source, .. } => &source.0.name,
+        }
+    }
+}
+
+impl FromStr for Package {
+    type Err = String;
+    /// "a#1" -> Package::Bucket
+    fn from_str(pkg: &str) -> Result<Self, Self::Err> {
+        let mut pkg_parts = pkg.split('#');
+        match (pkg_parts.next(), pkg_parts.next()) {
+            (Some(name), Some(bucket)) => Ok(Package::Bucket(Bucket {
+                name: name.to_string(),
+                bucket: bucket.parse().unwrap(),
+            })),
+            _ => Err(format!("{} is not a valid package name", pkg)),
+        }
+    }
+}
+
+impl Display for Package {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Package::Bucket(pkg) => write!(f, "{}", pkg),
+            Package::Proxi { source, target } => write!(f, "{}@{}->{}", source.0, source.1, target),
+        }
+    }
+}
+
+impl Display for Bucket {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}#{}", self.name, self.bucket)
+    }
+}
+
+impl Index {
+    /// List existing versions for a given package with newest versions first.
+    pub fn list_versions(&self, package: &Package) -> impl Iterator<Item = SemVer> + '_ {
+        match package {
+            // If we are on a bucket, we need to filter versions
+            // to only keep those within the bucket.
+            Package::Bucket(p) => {
+                let bucket_range = Range::between((p.bucket, 0, 0), (p.bucket + 1, 0, 0));
+                Either::Left(
+                    self.available_versions(&p.name)
+                        .filter(move |v| bucket_range.contains(*v))
+                        .cloned(),
+                )
+            }
+            // If we are on a proxi, there is one version per bucket in the target package.
+            // We can additionally filter versions to only those inside the dependency range.
+            Package::Proxi { target, source } => {
+                let dep_range = self
+                    .packages
+                    .get(&source.0.name)
+                    .unwrap()
+                    .get(&source.1)
+                    .unwrap()
+                    .get(target)
+                    .unwrap();
+                Either::Right(bucket_versions(
+                    self.available_versions(&target)
+                        .filter(move |v| dep_range.contains(v))
+                        .cloned(),
+                ))
+            }
+        }
+    }
+}
+
+/// Take a list of versions, and output a list of the corresponding bucket versions.
+/// So [1.1, 1.2, 2.3] -> [1.0, 2.0]
+fn bucket_versions(versions: impl Iterator<Item = SemVer>) -> impl Iterator<Item = SemVer> {
+    let mut current_bucket = None;
+    // This filter_map makes the hypothesis that versions are sorted in a normal or reverse order.
+    // Would need a bit more work if they are not ordered due to prioritizations, etc.
```

Comment:
This is a quite simple solution indeed. I wonder if we should lean towards "valid here but may bring problems for people using it as base" or "valid in most cases but a bit wasteful in resources". It might be safer to lean to your solution as it serves as a demonstration repo?